### ISSUE TITLE
Provide a stable Raiden Demo environment - Defining a plan

### ISSUE THREAD

**Issue by Dominik1999** [raiden-network/raiden#5187]
# Goal
The goal is to have a stable Raiden Demo environment, defined as a set of Raiden components, that consistently works every time we want to show Raiden to external people. 

Raiden components are defined as 
- the Raiden Client
- the Raiden Services
- the Matrix Server
- the Raiden Contracts
- the Raiden Wizard

We want someone to download the Raiden Wizard and it should work. It does not need to contain the latest features. Stability should not be affected by deploying new contracts, changes in the internal matrix server setup or any other change in the Raiden codebase.

In other words, we need a set of old components that work together and that we can run isolated somehow. 

# First things to consider
1. Find out which versions of the single components could work together
2. Find out if running on a private chain might be even more stable
3. For the services, we can deploy XYZ-demo as we did for testing
4. We need to come up with a plan on how to set up this demo environment until Mon 04th of November noon

**Comment by Dominik1999** [comment]
@christianbrb can you tell us the requirement about "how long should that demo environment work/ be online"?

After that answer:
 
@pirapira @lullis can you comment here in the issue and post the plans you had in mind? 

@karlb @pirapira @lullis we meet again Monday morning to define the plan

**Comment by pirapira** [comment]
* the goal should be, anybody performing the demo can press a button any time and it works.
    * this requires a fixed version of the whole codebase (or a docker image)
    * this also requires working servers (PFS & Matrix & Ethereum node)
        * I believe a self-contained local private-network setting would be much less prone to environmental changes (oh now the server is running a different version.)
* the requirements are similar to what @christianbrb wants often ("a version that can be tested").  So whatever @christianbrb can test successfully might become the "demo environtment".

**Comment by Dominik1999** [comment]
So we could say, we want the demo environment for 3 month before we update it again?

**Comment by lullis** [comment]
@pirapira, my understanding is that "demo environment" means more than "something that Christian can run on his laptop and show to other people". Rather, it needs to be a set of services that users can access and interact with anytime they want, only without the expectation of it being a production deployment. 

My suggestion is simply to take what in our subsystems are protocol/version-defining:

 - Contracts
 - Raiden Protocol, i.e, the specification of message formats and message sequence diagrams.
 - (Anything else I am missing?)

And then establish that all of our components should indicate what protocol/version they are complying to.

To illustrate, let's assume that the raiden contracts _v.X.Y.Z_ and a spec _ABC_ is the one that the different teams are testing and being planned to be demoed. Our components would then define their targets:

 - Raiden Client "version awesome-name-goes-here" (This release is to be used with raiden-contracts v.X.Y.Z and conforms to the spec ABC)
 - Raiden Services will be deployed at the url following the scheme https://pfs.<network>.raiden.network/<contract version>/<protocol version>. E.g, https://pfs.goerli.raiden.network/vX.Y.Z/ABC
 - Matrix Servers, AFAIK, do not depend on neither contract version or raiden protocol, so they just need to be deployed.
 - Raiden contracts (Does it depend on the spec version in any way, or is it self defining?)
 - Raiden Wizard (will search for download of raiden client release that uses contracts vX.Y.Z and complies to spec ABC, and will write the configuration file for the corresponding PFS version and contract addresses.

With the method above, we will have some increased complexity with deploying the raiden services,but the trade-off that it gets very easy to determine when an environment is demoable - all we have to do is ask "have we released/deployed all of the components of version XYZ and spec ABC?". It also eliminates the need to worry for "how long can we keep it online", because with versioned URLs/deployments we eliminate the conflict over URLs and allow to have *all* services running for as long as we might want/need them.

**Comment by christianbrb** [comment]
Thanks for your questions 😊

The main goal for the demo environment is, that
1. A user can download the Wizard (nothing more) and run Raiden using Infura
2. This needs to be always doable on the latest stable versions (tested and released)

@lullis explained it in a different way in his first paragraph of the comment above

**Comment by Dominik1999** [comment]
Ok, thanks for the clarification. I updated the goal above. What I understand now is that we need a "Wizard Testnet Environment" 

So we want that everyone can download the testnet Wizard and it simply works fine.

**Comment by karlb** [comment]
Looking at the services we need up to four simultaneous deployments to satisfy all requirements:
* Services master branch for the nightly scenario player runs
* One compatible with the latest Raiden RC for the testers
* One compatible with the current Raiden stable release for the testnet Wizard
* One compatible with the previous Raiden stable release, so that people won't have broken installations as soon as a new version is released

The first one can always have the same name `*.services-dev.raiden.network` across all releases, since we don't need to keep any compatibility there. It even *should* have the same name across releases so that we don't have to update the scenarios with constantly changing PFS addresses.
The other three should have one versioned name per release. That avoids breaking the setup when promoting released from RC to stable or when releasing a new stable.

I'm not convinced of the CONTRACT_VERSION/SPEC_VERSION path for versioning, yet. 
* Patch level versions of the contracts package don't change the contracts and are most likely compatible
* Do we have any reliable SPEC_VERSION yet?
* Spec changes will not always change the service messages, so we might need more service releases just to update the SPEC_VERSION without any other changes.
* Putting all versions in the path might cause ops problems when distributing the deployments over different machines. This can be solved by an additional reverse proxy, but it would introduce additional complexity.

Maybe we should just use the services version number and describe the compatibility in the release notes for now? We could also put it into the host name instead of the path, if that's better form an ops perspective (I'm not sure about that).

We should also consider not having all these deployments for all chains to keep the number of deployments manageable.

**Comment by lullis** [comment]
@rakanalh mentioned that we current do not have a way to version our protocol and that it does have any immediate reason to change frequently. In practice this means that the only component that is constantly varying and needs to be agreed upon for all systems to function correctly are the raiden contracts. These are already properly versioned.

With that in mind, the following plan would be established to get stable environments:

1. PFS deployments should be done to a canonical URL that includes contract version that it is using/depend on. E.g, instead of https://pfs-goerli.services-dev.raiden.network/, we should deploy https://pfs-goerli.services.raiden.network/v0.33.0.

1. Similar to the way that we keep an URL with the list of known matrix servers, we publish a document with a simple JSON mapping of raiden releases to contract versions.

1. The "-dev" and "-test" endpoints should only be treated as alias to whatever latest version is desired by the developers/integrators/testers.


@karlb, I do agree with the extra ops work, so this might be some work for Nils. However, the benefits would be substantial. With the points above, the wizard should then be able to determine (a) generate configuration files with the canonical URL and (b) know what is the most up-to-date release that is compatible and that can be downloaded. It wouldn't affect the existing setup and it would give us flexibility in turning the services on or off without being worried about conflicting with other systems and breaking expectations of other users.

**Comment by palango** [comment]
A few questions from my side:

> This needs to be always doable on the latest stable versions (tested and released)

Should this work on the testnet(s) and/or mainnet, if testnet which?

>  Rather, it needs to be a set of services that users can access and interact with anytime they want, with the only guarantee that the services will feature functionalities that are not necessarily the same that are on any production release.

Why would we show any other release other than the latest stable release?

What about just setting up a instance of the Raiden service bundle on agiven domain and use that for the demo and update it when we release a new raiden version?

**Comment by Dominik1999** [comment]
> > This needs to be always doable on the latest stable versions (tested and released)
> 
> Should this work on the testnet(s) and/or mainnet, if testnet which?

This should work on Görli

> Why would we show any other release other than the latest stable release?
> What about just setting up a instance of the Raiden service bundle on agiven domain and use that for the demo and update it when we release a new raiden version?

That would also work, we just do not have a latest stable release. For the purposes of the demo environment, we only want to have a stable environment that does not change when we introduce new bugs/features/contracts in any of the modules. So it must be a isolated set of Raiden Modules. Does that make sense?

**Comment by palango** [comment]
> So it must be a isolated set of Raiden Modules. Does that make sense?

Kind of. So currenty we want a stable environment for the wizard. The service bundle defaults to the services 0.4.0: https://github.com/raiden-network/raiden-transport/blob/master/docker-compose.yml#L4

So setting this up somewhere and using the RC1 raiden client should be stable as far as I understand? Or am I missing something?

**Comment by Dominik1999** [comment]
I think at the moment we don't have a sub-set of Raiden components - Client, Contracts, Services, Matrix, Wizard - that is stable. 

Goal of this issue is to define a plan, how we can achieve this. 

The "User-Story" is: 

Franz wants to be able to download the Wizard, and it works in a version of Raiden all the time. It is not important that this version is the latest.

**Comment by Dominik1999** [comment]
> So setting this up somewhere and using the RC1 raiden client should be stable as far as I understand? Or am I missing something?

Is this stable? RC1, Services 0.4.0, Matrix-Server (version?, how many?), contracts? 

I would understand as stable: The Wizard opens the WebUI and mediated payments to a random online node are possible.

**Comment by palango** [comment]
> Is this stable? RC1, Services 0.4.0, Matrix-Server (version?, how many?), contracts?

I think so.

> I would understand as stable: The Wizard opens the WebUI and mediated payments to a random online node are possible.

It would work for all random online nodes on the same matrix server.

**Comment by Dominik1999** [comment]
Ok if we already have a sub-set of Raiden components that is stable, we can use that for a demo environment.

**Comment by Dominik1999** [comment]
during my 2 days of sickness, we are 2 days late. 

Our draft plan for the demo environment:
- [ ] We need to have a demo “service bundle” environment deployed (PFS, MS, Matrix)
- [ ] We change the Wizard to hardcode the version of Raiden client to X (whenever we have a stable version X)
- [ ] Wizard runs, downloads Raiden and configures it to use the service bundle that is compatible with version X.

Currently, we are discussing whether we want to have different servers on which different versions of the service bundle are deployed.
- [ ] Option 1: We have 1 bundle (demo.raiden.network) deployed = 1 server, 1 domain. This is less maintainance effort and we only upgrade that if we release a breaking change that requires an upgrade. Means that we can only support 1 version of the Wizard because the wizard is going to be using a specific version of the contracts.
- [ ] Option 2: We deploy multiple versions on multiple servers and have multiple domains. This is more maintenance effort but means that we could support multiple versions of the Wizard.
